contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,479 @@
+through the same mechanism as OIRs in his Pod Spec.
+
+# Upgrading your cluster
+
+*TLDR:* 
+Given that we cannot guarantee that the Device Plugins are not running
+a daemon providing a critical service to Devices and when stopped will
+crash the running containers, it is up to the vendor to specify the
+upgrading scheme of their device plugin.
+
+However, If you are upgrading either Kubelet or any device plugin the safest way
+is to drain the node of all pods and upgrade.
+
+Depending on what you are upgrading and what changes happened then it
+is completely possible to only restart just Kubelet or just the device plugin.
+
+## Upgrading Kubelet
+
+This assumes that the Device Plugins running on the nodes fully implement the
+protocol and are able to recover from a Kubelet crash.
+
+Then, as long as the Device Plugin API does not change upgrading Kubelet can be done
+seamlessly through a Kubelet restart.
+
+*Currently:*
+As mentioned in the Versioning section, we currently expect the Device Plugin's
+API version to match exactly the Kubelet's Device Plugin API version.
+Therefore if the Device Plugin API version change then you will have to change
+the Device Plugin too.
+
+
+*Future:*
+When the Device Plugin API becomes a stable feature, versionning should be
+backward compatible and even if Kubelet has a different Device Plugin API,
+
+it should not require a Device Plugin upgrade.
+
+Refer to the versioning section for versioning scheme compatibility.
+
+## Upgrading Device Plugins
+
+Because we cannot enforce what the different Device Plugins will do, we cannot
+say for certain that upgrading a device plugin will not crash any containers
+on the node.
+
+It is therefore up to the Device Plugin vendors to specify if the Device Plugins
+can be upgraded without impacting any running containers.
+
+As mentioned earlier, the safest way is to drain the node before upgrading
+the Device Plugins.
+
+# Installation
+
+The installation process should be straightforward to the user, transparent
+and similar to other regular Kubernetes actions.
+The device plugin should also run in containers so that Kubernetes can
+deploy them and restart the plugins when they fail.
+However, we should not prevent the user from deploying a bare metal device
+plugin.
+
+Deploying the device plugins through DemonSets makes sense as the cluster
+admin would be able to specify which machines it wants the device plugins to
+run on, the process is similar to any Kubernetes action and does not require
+to change any parts of Kubernetes.
+
+Additionally, for integrated solutions such as `kubeadm` we can add support
+to auto-deploy community vetted Device Plugins.
+Thus not fragmenting once more the Kubernetes ecosystem.
+
+For users installing Kubernetes without using an integrated solution such
+as `kubeadm` they would use the examples that we would provide at:
+`https://github.com/Kubernetes/Kubernetes/tree/master/examples/device-plugin.yaml`
```

Comment:
This was discussed in the sig-node meeting and with @vishh.
I think we agreed that device plugins are provided by vendors and not Kubernetes.

So for Nvidia for example it would be something like: https://github.com/NVIDIA/Kubernetes-device-plugin/tree/master/examples/device-plugin.yaml`